On Windows the debug file is not deleted

In this file: delve\cmd\dlv\main.go

At line 101 this code creates the debug file:

```
 const debugname = "debug"
 goBuild := exec.Command("go", "build", "-o", debugname, "-gcflags", "-N -l", BuildFlags)
```

Then at line 113 in that same file this code does a deferred delete of that file:

```
 defer os.Remove(fp)
```

That code does not appear to be working on Windows and here is a simple example.

Initially there is no debug file:

```
C:\Go\src\github.com\jussij\DelveTest>dir
 Volume in drive C is SSD
 Volume Serial Number is 1020-B094

 Directory of C:\Go\src\github.com\jussij\DelveTest

02/02/2016  03:06 PM    <DIR>          .
02/02/2016  03:06 PM    <DIR>          ..
02/02/2016  02:50 PM         2,499,584 main.exe
31/01/2016  05:23 PM               562 main.go
               2 File(s)      2,500,146 bytes
```

Next, just start Delve and exit the debugger:

```
C:\Go\src\github.com\jussij\DelveTest>dlv debug main.go
Type 'help' for list of commands.
(dlv) exit
```

Now the debug file exists:

```
C:\Go\src\github.com\jussij\DelveTest>dir
 Volume in drive C is SSD
 Volume Serial Number is 1020-B094

 Directory of C:\Go\src\github.com\jussij\DelveTest

02/02/2016  03:03 PM    <DIR>          .
02/02/2016  03:03 PM    <DIR>          ..
02/02/2016  03:03 PM         2,500,096 debug
02/02/2016  02:50 PM         2,499,584 main.exe
31/01/2016  05:23 PM               562 main.go
               3 File(s)      5,000,242 bytes
```

So the debug file was not deleted.

> TestBuild in cmd/dlv/dlv_test.go could do the check at the end.

Thank you. TestBuild makes dlv "continue". But I would also like to test dlv "exit" before it runs program (see my example above).

Also TestBuild fails here:
```
c:\Users\Alex\dev\src\github.com\derekparker\delve\cmd\dlv>go test -v -run=Build github.com/derekparker/delve/cmd/dlv
=== RUN   TestBuild
--- FAIL: TestBuild (0.01s)
        dlv_test.go:62: makefile error make (c:\users\alex\dev\src\github.com\derekparker\delve): exec: "make": executable file not found in %PATH%
        dlv_test.go:63: output
        dlv_test.go:62: makefile error mingw32-make (c:\users\alex\dev\src\github.com\derekparker\delve): exec: "mingw32-make": executable file not found in %PATH%
        dlv_test.go:63: output
        dlv_test.go:35: failed assertion at dlv_test.go:66: make - exec: "mingw32-make": executable file not found in %PATH%
FAIL
exit status 1
FAIL    github.com/derekparker/delve/cmd/dlv    0.846s
```
because I do not have make. Why don't you use "go build" instead of "make"?

Alex

--- a/cmd/dlv/dlv_test.go
+++ b/cmd/dlv/dlv_test.go
@@ -135,6 +135,7 @@ func testIssue398(t *testing.T, dlvbin string, cmds []string) (stdout, stderr []
 	return
 }
 
+// TestIssue398 verifies that the debug executable is removed after exit.
 func TestIssue398(t *testing.T) {
 	tmpdir, err := ioutil.TempDir("", "TestIssue398")
 	if err != nil {

--- a/cmd/dlv/dlv_test.go
+++ b/cmd/dlv/dlv_test.go
@@ -2,7 +2,10 @@ package main
 
 import (
 	"bufio"
+	"bytes"
 	"flag"
+	"fmt"
+	"io/ioutil"
 	"os"
 	"os/exec"
 	"path/filepath"
@@ -95,4 +98,61 @@ func TestBuild(t *testing.T) {
 
 	client.Detach(true)
 	cmd.Wait()
+}
+
+func testIssue398(t *testing.T, dlvbin string, cmds []string) (stdout, stderr []byte) {
+	var stdoutBuf, stderrBuf bytes.Buffer
+	buildtestdir := filepath.Join(protest.FindFixturesDir(), "buildtest")
+	cmd := exec.Command(dlvbin, "debug")
+	cmd.Dir = buildtestdir
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		t.Fatal(err)
+	}
+	cmd.Stdout = &stdoutBuf
+	cmd.Stderr = &stderrBuf
+	if err := cmd.Start(); err != nil {
+		t.Fatal(err)
+	}
+	for _, c := range cmds {
+		fmt.Fprintf(stdin, "%s\n", c)
+	}
+	// ignore "dlv debug" command error, it returns
+	// errors even after successful debug session.
+	cmd.Wait()
+	stdout, stderr = stdoutBuf.Bytes(), stderrBuf.Bytes()
+
+	debugbin := filepath.Join(buildtestdir, "debug")
+	_, err = os.Stat(debugbin)
+	if err == nil {
+		t.Errorf("running %q: file %v was not deleted\nstdout is %q, stderr is %q", cmds, debugbin, stdout, stderr)
+		return
+	}
+	if !os.IsNotExist(err) {
+		t.Errorf("running %q: %v\nstdout is %q, stderr is %q", cmds, err, stdout, stderr)
+		return
+	}
+	return
+}
+
+func TestIssue398(t *testing.T) {
+	tmpdir, err := ioutil.TempDir("", "TestIssue398")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmpdir)
+
+	dlvbin := filepath.Join(tmpdir, "dlv.exe")
+	out, err := exec.Command("go", "build", "-o", dlvbin, "github.com/derekparker/delve/cmd/dlv").CombinedOutput()
+	if err != nil {
+		t.Fatalf("go build -o %v github.com/derekparker/delve/cmd/dlv: %v\n%s", dlvbin, err, string(out))
+	}
+
+	testIssue398(t, dlvbin, []string{"exit"})
+
+	const hello = "hello world!\n"
+	stdout, _ := testIssue398(t, dlvbin, []string{"continue", "exit"})
+	if !strings.Contains(string(stdout), hello) {
+		t.Errorf("stdout %q should contain %q", stdout, hello)
+	}
 }

--- a/cmd/dlv/dlv_test.go
+++ b/cmd/dlv/dlv_test.go
@@ -150,7 +150,7 @@ func TestIssue398(t *testing.T) {
 
 	testIssue398(t, dlvbin, []string{"exit"})
 
-	const hello = "hello world!\n"
+	const hello = "hello world!"
 	stdout, _ := testIssue398(t, dlvbin, []string{"continue", "exit"})
 	if !strings.Contains(string(stdout), hello) {
 		t.Errorf("stdout %q should contain %q", stdout, hello)

--- a/pkg/proc/native/proc.go
+++ b/pkg/proc/native/proc.go
@@ -4,7 +4,6 @@ import (
 	"errors"
 	"fmt"
 	"go/ast"
-	"os"
 	"runtime"
 	"sync"
 
@@ -14,9 +13,8 @@ import (
 // Process represents all of the information the debugger
 // is holding onto regarding the process we are debugging.
 type Process struct {
-	bi      proc.BinaryInfo
-	pid     int         // Process Pid
-	Process *os.Process // Pointer to process struct for the actual process we are debugging
+	bi  proc.BinaryInfo
+	pid int // Process Pid
 
 	// Breakpoint table, holds information on breakpoints.
 	// Maps instruction address to Breakpoint struct.
@@ -371,13 +369,7 @@ func (dbp *Process) FindBreakpoint(pc uint64) (*proc.Breakpoint, bool) {
 
 // Returns a new Process struct.
 func initializeDebugProcess(dbp *Process, path string) (*Process, error) {
-	process, err := os.FindProcess(dbp.pid)
-	if err != nil {
-		return nil, err
-	}
-
-	dbp.Process = process
-	err = dbp.LoadInformation(path)
+	err := dbp.LoadInformation(path)
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/proc/native/proc.go
+++ b/pkg/proc/native/proc.go
@@ -444,6 +444,7 @@ func (dbp *Process) postExit() {
 	dbp.exited = true
 	close(dbp.ptraceChan)
 	close(dbp.ptraceDoneChan)
+	dbp.bi.Close()
 }
 
 func (dbp *Process) writeSoftwareBreakpoint(thread *Thread, addr uint64) error {

--- a/pkg/proc/native/proc_windows.go
+++ b/pkg/proc/native/proc_windows.go
@@ -177,7 +177,10 @@ func (dbp *Process) Kill() error {
 	// but some tests appear to Kill twice causing
 	// this to fail on second attempt.
 	_ = syscall.TerminateProcess(dbp.os.hProcess, 1)
-	dbp.exited = true
+	dbp.execPtraceFunc(func() {
+		dbp.waitForDebugEvent(waitBlocking)
+	})
+	dbp.postExit()
 	return nil
 }
 

--- a/pkg/proc/native/proc_windows.go
+++ b/pkg/proc/native/proc_windows.go
@@ -173,13 +173,24 @@ func (dbp *Process) Kill() error {
 	if !dbp.threads[dbp.pid].Stopped() {
 		return errors.New("process must be stopped in order to kill it")
 	}
+
+	p, err := os.FindProcess(dbp.pid)
+	if err != nil {
+		return err
+	}
+	defer p.Release()
+
 	// TODO: Should not have to ignore failures here,
 	// but some tests appear to Kill twice causing
 	// this to fail on second attempt.
 	_ = syscall.TerminateProcess(dbp.os.hProcess, 1)
+
 	dbp.execPtraceFunc(func() {
 		dbp.waitForDebugEvent(waitBlocking)
 	})
+
+	p.Wait()
+
 	dbp.postExit()
 	return nil
 }

--- a/pkg/proc/native/proc_windows.go
+++ b/pkg/proc/native/proc_windows.go
@@ -473,5 +473,7 @@ func killProcess(pid int) error {
 	if err != nil {
 		return err
 	}
+	defer p.Release()
+
 	return p.Kill()
 }

